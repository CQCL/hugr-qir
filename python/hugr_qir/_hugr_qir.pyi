def cli(*args: str) -> None: ...
def opt_level_choices() -> list[str]: ...
def compile_target_choices() -> list[str]: ...
def output_format_choices() -> list[str]: ...
