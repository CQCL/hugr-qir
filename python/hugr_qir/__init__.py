def main_cli():
    from ._hugr_qir import cli
    import sys
    cli(*sys.argv)
