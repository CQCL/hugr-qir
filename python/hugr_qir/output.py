from base64 import b64encode
from enum import Enum
from pathlib import Path

from llvmlite.binding import (  # type: ignore[import-untyped]
    create_context,
    parse_assembly,
)


class OutputFormat(Enum):
    LLVMIR = "llvmir"
    BITCODE = "bitcode"
    BASE64 = "base64"


def get_write_mode(out_format: OutputFormat | None, file_path: Path | None) -> str:
    if out_format and out_format in [OutputFormat.BITCODE, OutputFormat.BASE64]:
        return "wb"
    if file_path and file_path.suffix not in [".ll", ".asm"]:
        return "wb"
    return "w"


def ir_string_to_output_format(qir_ir: str, output_format: OutputFormat) -> str | bytes:
    match output_format:
        case OutputFormat.LLVMIR:
            return qir_ir
        case OutputFormat.BITCODE:
            ctx = create_context()
            module = parse_assembly(qir_ir, context=ctx)
            return module.as_bitcode()
        case OutputFormat.BASE64:
            ctx = create_context()
            module = parse_assembly(qir_ir, context=ctx)
            qir_bitcode = module.as_bitcode()
            return b64encode(qir_bitcode).decode("utf-8")
    errmsg = "Unrecognized output format"
    raise ValueError(errmsg)
