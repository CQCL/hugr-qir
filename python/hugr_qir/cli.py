"""Cli for hugr-qir."""

import tempfile
from importlib.metadata import version
from pathlib import Path

import click
from quantinuum_qircheck import qircheck

from hugr_qir._hugr_qir import cli


@click.command(name="hugr-qir")
@click.argument("hugr_file", type=click.Path(exists=True, path_type=Path))
@click.option(
    "--validate-qir/--no-validate-qir",
    "validate",
    default=True,
    help="Whether to validate output",
)
@click.option(
    "-o",
    "--output",
    "outfile",
    type=click.File("w"),
    default="-",
    help="Name of output file (optional)",
)
@click.version_option(version=version("hugr_qir"))
def hugr_qir(validate: bool, hugr_file: Path, outfile: click.File) -> None:
    """This is the cli for converting hugr to qir."""

    options = ["-q"]
    with tempfile.NamedTemporaryFile(delete=True, suffix=".ll") as temp_file:
        tmp_options = [*options, "-o", temp_file.name]
        cli(str(hugr_file), *tmp_options)
        with Path.open(Path(temp_file.name)) as output:
            qir = output.read()
    if validate:
        qircheck(qir)

    if hasattr(outfile, "write"):
        outfile.write(qir)
    else:
        # this should never happen but the check for write is
        # needed to appease mypy (click doesn't handle typing well)
        err_msg = f"Could not write to object {outfile}"
        raise AttributeError(err_msg)


if __name__ == "__main__":
    hugr_qir()
