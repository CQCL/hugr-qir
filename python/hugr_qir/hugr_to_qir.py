import tempfile
from pathlib import Path

from .cli import hugr_qir_impl


def hugr_to_qir(hugr: bytes, validate_qir: bool = True) -> str:
    """A function for converting hugr to qir."""
    with (
        tempfile.NamedTemporaryFile(delete=True, suffix=".hugr") as temp_infile,
        tempfile.NamedTemporaryFile(delete=True, suffix=".ll") as temp_outfile,
    ):
        with Path.open(Path(temp_infile.name), "wb") as cli_input:
            cli_input.write(hugr)
        with Path.open(Path(temp_outfile.name), "w") as cli_output:
            hugr_qir_impl(validate_qir, Path(cli_input.name), cli_output)
        with Path.open(Path(temp_outfile.name), "r") as cli_output:
            return cli_output.read()
