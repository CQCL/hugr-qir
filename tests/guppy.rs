use std::{env, path::Path, process::{self, Command, Stdio}};

use hugr::{package::Package, std_extensions::STD_REG, Hugr};
use hugr_llvm::inkwell;
use hugr_qir::CompileArgs;
use insta::assert_snapshot;
use itertools::Itertools as _;
use rstest::rstest;

fn capture_guppy(path: impl AsRef<Path>) -> (Hugr, String) {
    let process::Output {
        status,
        stdout,
        stderr,
    } = Command::new(".devenv/state/venv/bin/python")
        .arg(path.as_ref())
        .stdin(Stdio::null())
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .output()
        .unwrap();
    eprintln!("{}", std::str::from_utf8(&stdout).unwrap());
    eprintln!("{}", std::str::from_utf8(&stderr).unwrap());
    assert!(status.success());
    let hugr = Package::from_json_reader(stdout.as_slice(), &STD_REG)
        .unwrap()
        .modules
        .into_iter()
        .exactly_one()
        .unwrap();
    let stderr = String::from_utf8(stderr).unwrap();
    (hugr, stderr)
}

fn compile(hugr: &mut Hugr) -> String {
    let args = {
        let mut a = CompileArgs::default();
        a.validate = true;
        a
    };
    let context = inkwell::context::Context::create();
    let module = args.compile(hugr, &context).unwrap();
    module.to_string()
}

// This generates a test case for each file matching 'guppy_examples/**/*.py'
//
// For now we snapshot test the stderr output of executing the python file,
// and the LLVM IR output by hugr-qir.
//
// If we get problems with new test cases not being run in dev environments we might consider
// adding a `build.rs` as in:
// <https://docs.rs/rstest/latest/rstest/attr.rstest.html#files-path-as-input-arguments>
#[rstest]
fn guppy_examples(
    #[base_dir = "guppy_examples"]
    #[files("**/*.py")]
    file: impl AsRef<Path>,
) {
    let file2 = file.as_ref();
    let mut settings = insta::Settings::clone_current();
    settings.set_snapshot_suffix(file2.file_stem().unwrap().to_str().unwrap());
    settings.bind(|| {
        let (mut hugr, stderr) = capture_guppy(file);
        assert_snapshot!("llvmir", compile(&mut hugr));
        assert_snapshot!("stderr", stderr);
    });
}
