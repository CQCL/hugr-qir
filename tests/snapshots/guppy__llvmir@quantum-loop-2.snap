---
source: tests/guppy.rs
expression: compile(&mut hugr)
---
; ModuleID = 'hugr-qir'
source_filename = "hugr-qir"

%QUBIT = type opaque
%RESULT = type opaque

@0 = private unnamed_addr constant [37 x i8] c"Expected variant 1 but got variant 0\00", align 1
@prelude.panic_template = private unnamed_addr constant [34 x i8] c"Program panicked (signal %i): %s\0A\00", align 1
@1 = private unnamed_addr constant [37 x i8] c"Expected variant 1 but got variant 0\00", align 1
@prelude.panic_template.1 = private unnamed_addr constant [34 x i8] c"Program panicked (signal %i): %s\0A\00", align 1
@2 = private unnamed_addr constant [2 x i8] c"1\00", align 1
@3 = private unnamed_addr constant [2 x i8] c"0\00", align 1
@4 = private unnamed_addr constant [37 x i8] c"Expected variant 1 but got variant 0\00", align 1
@prelude.panic_template.2 = private unnamed_addr constant [34 x i8] c"Program panicked (signal %i): %s\0A\00", align 1

define void @__hugr__.main.1() #0 {
alloca_block:
  %0 = call %QUBIT* @__quantum__rt__qubit_allocate()
  %1 = insertvalue { i1, %QUBIT* } { i1 true, %QUBIT* poison }, %QUBIT* %0, 1
  %2 = call %QUBIT* @__quantum__rt__qubit_allocate()
  %3 = insertvalue { i1, %QUBIT* } { i1 true, %QUBIT* poison }, %QUBIT* %2, 1
  %4 = extractvalue { i1, %QUBIT* } %1, 0
  %cond = icmp eq i1 %4, true
  br i1 %cond, label %cond_133_case_1, label %cond_133_case_0

cond_133_case_1:                                  ; preds = %alloca_block
  %5 = extractvalue { i1, %QUBIT* } %1, 1
  br label %cond_exit_133

6:                                                ; preds = %26, %cond_exit_122
  %"19_1.0" = phi i64 [ %38, %cond_exit_122 ], [ %73, %26 ]
  %"19_0.0" = phi i64 [ %37, %cond_exit_122 ], [ %72, %26 ]
  %7 = insertvalue { i64, i64 } poison, i64 %"19_0.0", 0
  %8 = insertvalue { i64, i64 } %7, i64 %"19_1.0", 1
  %9 = extractvalue { i64, i64 } %8, 0
  %10 = extractvalue { i64, i64 } %8, 1
  %11 = insertvalue { i64, i64 } poison, i64 %9, 0
  %12 = insertvalue { i64, i64 } %11, i64 %10, 1
  %13 = icmp slt i64 %9, %10
  %14 = select i1 %13, i1 true, i1 false
  %15 = insertvalue { { i64, i64 }, i1 } poison, i1 %14, 1
  %16 = insertvalue { { i64, i64 }, i1 } %15, { i64, i64 } %12, 0
  %17 = extractvalue { { i64, i64 }, i1 } %16, 1
  %18 = extractvalue { { i64, i64 }, i1 } %16, 0
  %19 = insertvalue { { i64, i64 }, i1 } poison, i1 %17, 1
  %20 = insertvalue { { i64, i64 }, i1 } %19, { i64, i64 } %18, 0
  %21 = extractvalue { { i64, i64 }, i1 } %20, 1
  %22 = extractvalue { { i64, i64 }, i1 } %20, 0
  %23 = extractvalue { i64, i64 } %22, 0
  %24 = extractvalue { i64, i64 } %22, 1
  %cond225 = icmp eq i1 %21, true
  br i1 %cond225, label %48, label %39

cond_144_case_1:                                  ; preds = %48
  %25 = extractvalue { i1, %QUBIT* } %50, 1
  br label %cond_exit_144

26:                                               ; preds = %cond_exit_144, %75
  br label %6

cond_133_case_0:                                  ; preds = %alloca_block
  %27 = extractvalue { i32, i8* } { i32 1, i8* getelementptr inbounds ([37 x i8], [37 x i8]* @0, i32 0, i32 0) }, 0
  %28 = extractvalue { i32, i8* } { i32 1, i8* getelementptr inbounds ([37 x i8], [37 x i8]* @0, i32 0, i32 0) }, 1
  %29 = call i32 (i8*, ...) @printf(i8* getelementptr inbounds ([34 x i8], [34 x i8]* @prelude.panic_template, i32 0, i32 0), i32 %27, i8* %28)
  call void @abort()
  br label %cond_exit_133

cond_exit_133:                                    ; preds = %cond_133_case_1, %cond_133_case_0
  %"0.0" = phi %QUBIT* [ null, %cond_133_case_0 ], [ %5, %cond_133_case_1 ]
  %30 = extractvalue { i1, %QUBIT* } %3, 0
  %cond224 = icmp eq i1 %30, true
  br i1 %cond224, label %cond_122_case_1, label %cond_122_case_0

cond_122_case_1:                                  ; preds = %cond_exit_133
  %31 = extractvalue { i1, %QUBIT* } %3, 1
  br label %cond_exit_122

cond_122_case_0:                                  ; preds = %cond_exit_133
  %32 = extractvalue { i32, i8* } { i32 1, i8* getelementptr inbounds ([37 x i8], [37 x i8]* @1, i32 0, i32 0) }, 0
  %33 = extractvalue { i32, i8* } { i32 1, i8* getelementptr inbounds ([37 x i8], [37 x i8]* @1, i32 0, i32 0) }, 1
  %34 = call i32 (i8*, ...) @printf(i8* getelementptr inbounds ([34 x i8], [34 x i8]* @prelude.panic_template.1, i32 0, i32 0), i32 %32, i8* %33)
  call void @abort()
  br label %cond_exit_122

cond_exit_122:                                    ; preds = %cond_122_case_1, %cond_122_case_0
  %"09.0" = phi %QUBIT* [ null, %cond_122_case_0 ], [ %31, %cond_122_case_1 ]
  %35 = insertvalue { i64, i64 } poison, i64 0, 0
  %36 = insertvalue { i64, i64 } %35, i64 10, 1
  %37 = extractvalue { i64, i64 } %36, 0
  %38 = extractvalue { i64, i64 } %36, 1
  br label %6

39:                                               ; preds = %6
  %40 = call %RESULT* @__quantum__qis__m__body(%QUBIT* %"0.0")
  %41 = call i1 @__quantum__qis__read_result__body(%RESULT* %40)
  %42 = select i1 %41, i1 true, i1 false
  %43 = call %RESULT* @__quantum__qis__m__body(%QUBIT* %"09.0")
  %44 = call i1 @__quantum__qis__read_result__body(%RESULT* %43)
  %45 = select i1 %44, i1 true, i1 false
  %46 = select i1 %42, i1 true, i1 false
  call void @__quantum__rt__bool_record_output(i1 %46, i8* getelementptr inbounds ([2 x i8], [2 x i8]* @2, i32 0, i32 0))
  %47 = select i1 %45, i1 true, i1 false
  call void @__quantum__rt__bool_record_output(i1 %47, i8* getelementptr inbounds ([2 x i8], [2 x i8]* @3, i32 0, i32 0))
  ret void

48:                                               ; preds = %6
  %49 = call %QUBIT* @__quantum__rt__qubit_allocate()
  %50 = insertvalue { i1, %QUBIT* } { i1 true, %QUBIT* poison }, %QUBIT* %49, 1
  %51 = extractvalue { i1, %QUBIT* } %50, 0
  %cond226 = icmp eq i1 %51, true
  br i1 %cond226, label %cond_144_case_1, label %cond_144_case_0

cond_144_case_0:                                  ; preds = %48
  %52 = extractvalue { i32, i8* } { i32 1, i8* getelementptr inbounds ([37 x i8], [37 x i8]* @4, i32 0, i32 0) }, 0
  %53 = extractvalue { i32, i8* } { i32 1, i8* getelementptr inbounds ([37 x i8], [37 x i8]* @4, i32 0, i32 0) }, 1
  %54 = call i32 (i8*, ...) @printf(i8* getelementptr inbounds ([34 x i8], [34 x i8]* @prelude.panic_template.2, i32 0, i32 0), i32 %52, i8* %53)
  call void @abort()
  br label %cond_exit_144

cond_exit_144:                                    ; preds = %cond_144_case_1, %cond_144_case_0
  %"0110.0" = phi %QUBIT* [ null, %cond_144_case_0 ], [ %25, %cond_144_case_1 ]
  call void @__quantum__qis__phasedx__body(double 0x3FF921FB54442D18, double 0xBFF921FB54442D18, %QUBIT* %"0110.0")
  call void @__quantum__qis__rz__body(double 0x400921FB54442D18, %QUBIT* %"0110.0")
  %55 = call %RESULT* @__quantum__qis__m__body(%QUBIT* %"0110.0")
  %56 = call i1 @__quantum__qis__read_result__body(%RESULT* %55)
  %57 = select i1 %56, i1 true, i1 false
  %58 = insertvalue { i64, i64 } poison, i64 %23, 0
  %59 = insertvalue { i64, i64 } %58, i64 %24, 1
  %60 = extractvalue { i64, i64 } %59, 0
  %61 = extractvalue { i64, i64 } %59, 1
  %62 = add i64 %60, 1
  %63 = insertvalue { i64, i64 } poison, i64 %62, 0
  %64 = insertvalue { i64, i64 } %63, i64 %61, 1
  %65 = insertvalue { { i64, i64 }, i64 } poison, i64 %60, 1
  %66 = insertvalue { { i64, i64 }, i64 } %65, { i64, i64 } %64, 0
  %67 = extractvalue { { i64, i64 }, i64 } %66, 1
  %68 = extractvalue { { i64, i64 }, i64 } %66, 0
  %69 = insertvalue { { i64, i64 }, i64 } poison, i64 %67, 1
  %70 = insertvalue { { i64, i64 }, i64 } %69, { i64, i64 } %68, 0
  %71 = extractvalue { { i64, i64 }, i64 } %70, 0
  %72 = extractvalue { i64, i64 } %71, 0
  %73 = extractvalue { i64, i64 } %71, 1
  %74 = select i1 %57, i1 true, i1 false
  %cond227 = icmp eq i1 %74, true
  br i1 %cond227, label %75, label %26

75:                                               ; preds = %cond_exit_144
  call void @__quantum__qis__phasedx__body(double 0x3FF921FB54442D18, double 0xBFF921FB54442D18, %QUBIT* %"09.0")
  call void @__quantum__qis__rz__body(double 0x400921FB54442D18, %QUBIT* %"09.0")
  br label %26
}

declare %QUBIT* @__quantum__rt__qubit_allocate()

declare i32 @printf(i8*, ...)

declare void @abort()

declare %RESULT* @__quantum__qis__m__body(%QUBIT*)

declare i1 @__quantum__qis__read_result__body(%RESULT*)

declare void @__quantum__rt__bool_record_output(i1, i8*)

declare void @__quantum__qis__phasedx__body(double, double, %QUBIT*)

declare void @__quantum__qis__rz__body(double, %QUBIT*)

attributes #0 = { "entry_point" "output_labeling_schema" "qir_profiles"="custom" "required_num_qubits"="20" "required_num_results"="20" }
